### ISSUE TITLE
Add T-matrix calculations for non-spherical scatterers

### ISSUE THREAD

**Issue by vnmanoharan** [manoharan-lab/holopy#148]
Anna had written a wrapper around Michael Mishchenko's Fortran code for axisymmetric scatterers, but it was too out of sync with the current codebase to merge (see #7).  I'm filing this issue so that we don't forget about this.

**Comment by LaurentRDC** [comment]
Hi all,

My group and I are interested in getting this done. I'm not familiar with Holopy, so any pointers as to what is different now from then is appreciated.

**Comment by anna-wang** [comment]
Hi Laurent - thanks for your interest! Are you hoping to calculate
holograms, or just get the scattered field? If it's the latter, looking at
the wrapper I wrote all those years ago will give you an idea of the input
Mishchenko's code requires. All the changes I made are in the headers of
the Fortran files.

At the moment I can't comment on how HoloPy has changed, but I'll have more
time in a month or so to start looking into this.

Anna

**Comment by LaurentRDC** [comment]
We are looking to fit holograms to scattering from, say, a cylinder. I'll play around and report back.

**Comment by barkls** [comment]
Hi Laurent,

I am the current maintainer of HoloPy, so I am up to date on its current status. Please don't hesitate to reach out with any questions you may have.

We have just finished a release of HoloPy that was not backwards compatible, so that's why Anna's code needs a bit of work to make it usable. The best way to get a feel for the current status of HoloPy is to go through the tutorials in our [user guide](https://holopy.readthedocs.io). We don't quite have the conda installation working yet, so you will need to follow the installation instructions on the [developer's page](http://holopy.readthedocs.io/en/master/users/dev_tutorial.html). 

I forked Anna's code and did a preliminary pass on [updating it](https://github.com/barkls/holopy/tree/tmatrix). I removed a bunch of extraneous files, fixed merge conflicts, and got it to work to the point of being able to run setup.py and import holopy with python 3. Some work will be required to get the T-matrix code to actually work though. 

The biggest changes in our release were how we pass information around within HoloPy. For example, we formerly had an ``Optics`` class that contained information about wavelength, refractive index, etc. of an experimental setup. These values are now included as attributes within xarray DataArray objects. Essentially, Anna's code amounts to adding a scattering theory to HoloPy. There are some pointers in our developer's guide (linked above), but the best resource is probably to compare to one of the existing HoloPy scattering theories, such as in holopy/scattering/theory/mie.py. 

I should also point out that HoloPy can use the ADDA package to model arbitrary shapes. This code currently works, so it would be easier to get going (although I have never used it myself. Anna is a better reference on that). However, it is much more computationally inefficient, so it depends what your priorities are.

**Comment by vnmanoharan** [comment]
@barkls, it looks like there are still some binary files (.o and .exe) lying around.  Can you make a new branch just by adding the source files, but without any of the revision history from the previous branch?  That will keep the repository clean (having the .exe and .o files in the revision history bloats the repo, regardless of whether they're in the current working tree). 

@anna-wang, do we need both classes (double and extended precision)? It looks like there is a lot of redundant code.  Could we get away with simply using one of them, for example the extended precision code?

@LaurentRDC, @barkls, @anna-wang, because this module calls the fortran executable and parses the output rather than linking to a fortran library, it shouldn't be too difficult to get it integrated. From reading the [development manual](http://holopy.readthedocs.io/en/master/users/dev_tutorial.html#adding-a-new-scattering-theory), it seems we'll need to convert the existing `_calc_field()` method in tmatrix.py to either a method `_raw_scat_matrs()` that returns the raw scattering matrices or `_raw_fields()` that returns the raw fields. @barkls, please confirm.

**Comment by barkls** [comment]
Yes, I intentionally left those files for now. Without a functioning setup.py, it might be difficult to build them from source and Anna was at least able to use the .exe to get code to run previously. If it's easy to reproduce them then I agree they should be removed from the repository. 

From my cursory look earlier, it appears fairly straightforward to get this code functioning in the most recent HoloPy. Anna has already done the hard part of interfacing with the Fortran code. There are additions to code in a few unexpected places that probably don't belong. If they are necessary for the current implementation of T-matrix to work then some retooling will be required, but they might just be there for testing purposes.

**Comment by vnmanoharan** [comment]
It's trivial to compile the code and produce the .o and .exe files -- it's one line to be typed at the command prompt. No setup.py is needed, at least for now.

**Comment by vnmanoharan** [comment]
Also, it's not a matter of removing the binaries from the repo, because they are still in the revision history and will unnecessarily bloat the repository. So I'd like us to have a new branch with just the source files added, and we can work from there.

**Comment by vnmanoharan** [comment]
Confirmed that code compiles with gfortran using the following line:
```
gfortran -o S.exe S.lp.f lpd.f
```
for double precision code, 
```
gfortran -o S.exe S.lp.f lpq.f
```
for extended precision. Compiled on Windows 10 with gfortran installed from anaconda.

**Comment by anna-wang** [comment]
My memory is terrible, but I think I used the double precision code because
I couldn't compile the extended precision code with gfortran at the time (I
had to use Intel's fortran compiler). The double precision code is fine
until the aspect ratio of the particles gets larger than ~5 for particles
with size parameter ~5, then it has convergence issues.

I also think the extended precision code is slower but I'm not sure by how
much.

Anna

**Comment by barkls** [comment]
I've updated [my branch](https://github.com/barkls/holopy/tree/tmatrix) to include only the raw T-matrix code files, added to HoloPy3 as a single commit.

**Comment by vnmanoharan** [comment]
@LaurentRDC, @barkls, @anna-wang:
I've made some fixes to the code, which is available in [my branch](https://github.com/vnmanoharan/holopy/tree/tmatrix). Unfortunately I can't test it, since `mieangfuncs` cannot yet be compiled on windows. Can someone test the code on Mac or Linux to see if it works? Here are some instructions:

1) `git clone` my repository and `git checkout` the tmatrix branch
2) cd to the repository root, [REPO_ROOT], and compile the extension with 
```
python setup.py build_ext --inplace
```
3) cd to [REPO_ROOT]/holopy/scattering/theory/tmatrix_f
4) build the T-matrix executable using
```
gfortran -o S.exe S.lp.f lpq.f
```
5) cd back up to [REPO_ROOT]
6) in an iPython shell or notebook, do
```
%run holopy/scattering/theory/example_tmatrix.py
```
7) then look at `holo_spheroid` and see if it matches [this image](https://github.com/anna-wang/holopy/blob/98470353c8e429c45b54fdbd814af004241c1666/holopy/scattering/theory/example_tmatrix.png).

**Comment by jeromefung** [comment]
holo_spheroid doesn't match -- but it looks like a scatterer coordinate
issue to me. See the attached.

Compilation proceeded with no errors, but I did get the following warning
when running example_tmatrix.py:

Note: The following floating-point exceptions are signalling:
IEEE_UNDERFLOW_FLAG IEEE_DENORMAL

This was carried out on a Mac running OS 10.10.5, Python 3.5.3, and
gfortran 4.9.3.

Jerome

**Comment by vnmanoharan** [comment]
@jeromefung, thanks for trying it. The attachment doesn't seem to have come through, though.  Can you try adding it directly to a new post on the [issue page](https://github.com/manoharan-lab/holopy/issues/148)?

**Comment by jeromefung** [comment]
@vnmanoharan, here's the attachment.

![holo_spheroid](https://user-images.githubusercontent.com/7784176/27957742-c70053ce-62ed-11e7-8649-0d6fd6891a81.png)

**Comment by anna-wang** [comment]
Hi everyone,

I get what Jerome got, but looking at example_tmatrix.py I think the
original image that we're comparing it against is wrong.

The COM coordinates are (10,10,20)
The pixel spacing is 0.1/pixel
The image width is 100 pixels
So in pixels, the COM should be at (100,100)

i.e. the COM should be the bottom right corner, which it is.

Anna

**Comment by vnmanoharan** [comment]
@anna-wang, @jeromefung: looks like I put the wrong detector size in the example script.  The original image was 200x200, and the current one is 100x100.  I've pushed a fix to my branch.  Can you fetch the changes and try again?  Or just change the line in `example_tmatrix.py` that reads
```
detector = hp.detector_grid(shape = 100, spacing = .1)
```
to
```
detector = hp.detector_grid(shape = 200, spacing = .1)
```

**Comment by jeromefung** [comment]
With the new pull, my `holo_spheroid` matches (visually) @anna-wang 's original example.

**Comment by anna-wang** [comment]
Yes it agrees for me now too. I forgot to mention earlier that I too get
the warning Jerome gets:

Note: The following floating-point exceptions are signalling:
IEEE_UNDERFLOW_FLAG IEEE_DENORMAL

**Comment by vnmanoharan** [comment]
I get that error too, but at least it seems to be working now.  Thanks for the help.  I still need to clean up the scatterer classes.  Each option for the axisymmetric shapes in the T-matrix code should be mapped onto its own class.  Also need to figure out how to get the executable to build using setup.py.

**Comment by barkls** [comment]
Right now we have a few Scatterer classes that overlap. 

Axisymmetric is a general class for Tmatrix calculations.

Cylinder could be a subclass of axisymmetric that specifically has shape =  -2.

Spheroid could be a subclass of axisymmetric that specifically has shape =  -1.

This would simplify the machinery for tmatrix calculations since all compatible scatterers would be within the Axisymmetric class. I can implement these changes if @vnmanoharan is in agreement.

**Comment by barkls** [comment]
A second option is to have Cylinder and Spheroid be their own classes, but then have ``Tmatrix`` instantiate a general ``Axisymmetric`` object based on which class of scatterer is passed in. This has the advantage that it would also be able to handle spheres. 

A third option would be to attach methods to the appropriate scatterer classes so that they know how to transform themselves into an ``Axisymmetric`` object for tmatrix calculations.

**Comment by vnmanoharan** [comment]
I would like to keep scatterers as independent of theories as possible.  In other words, we should define scatterers based on their shape, not necessarily based on which theory can handle them. That makes it easier to switch between theories for the same scatterer. For example, spheroids can be handled by either DDA or T-matrix.  Spheres can be handled by either Mie, DDA, T-matrix, or Multisphere.

So I propose to get rid of the Axisymmetric class and the shape attribute entirely, and just define Spheroid, and Cylinder classes. Later, if needed, we can define Chebyshev and generalized Chebyshev scatterer classes.  But for now I think we just need Spheroid and Cylinder.

**Comment by barkls** [comment]
That sounds like a good plan for scatterer classes. I will implement it.

For compiling the executable, it looks fairly straightforward to call gfortran from within a ``setup.py`` using ``subprocess``. (see http://www.legi.grenoble-inp.fr/people/Pierre.Augier/how-to-call-fortran-from-python.html or https://stackoverflow.com/questions/39988621/running-fortran-executable-within-python-script/39988971#39988971). Things would get more complicated if we want to be able to use different compilers.

**Comment by vnmanoharan** [comment]
Yes, we can use `subprocess` to call `gfortran` from setup.py, as shown in the StackOverflow page you linked to. I'm just not sure that is the right way to do it.  Usually when one wants to distribute a package for compilation on multiple platforms, one would use [GNU autoconf](http://www.gnu.org/software/autoconf/autoconf.html) to make a configuration script, which will automatically determine what compilers are available on the platform and then generate the appropriate makefile. I don't know how to do that with distutils.  

It might be possible, however, to use both autoconf, make, and distutils together:
https://stackoverflow.com/questions/33168482/compiling-installing-c-executable-using-pythons-setuptools-setup-py

**Comment by barkls** [comment]
I agree using subprocess to call gfortran is not elegant. Maybe it's possible to call autoconf and make with subprocess?

I think it's also worth considering the different installation situations we have currently:

For built distributions, we are already compiling fortran using gcc on conda-forge (mac/linux) or ifort on @ralex0 's Windows computer. Users do not need to ever run the setup.py or choose a Fortran compiler. It might be ok in these cases to call subprocess on gfortran or ifort directly from the setup.py since these are controlled environments that will not have different compilers available.

For developer installations, we could require a separate compile step in addition to running setup.py if it's not easy to figure out a flexible way to combine them.

**Comment by vnmanoharan** [comment]
For python modules compiled from fortran, f2py and numpy.distutils automatically figure out which fortran compiler to use (https://docs.scipy.org/doc/numpy-dev/f2py/distutils.html).  I haven't found a way to use these same tools to compile an executable, though. The solution is probably for us to generate a configure script (using autoconf) that we would include in the repository. Then, in setup.py, we call that script, and call `make` on the resulting makefile, to compile the executable.  This is the approach that was used in the StackOverflow article above:
https://stackoverflow.com/questions/33168482/compiling-installing-c-executable-using-pythons-setuptools-setup-py

If I understand correctly, we would actually have to provide two configure scripts: one that works on platforms where bash is available (Linux, MacOS), and one (call it configure.bat) that works on Windows.

**Comment by vnmanoharan** [comment]
Actually, I'm going to propose something far simpler. Let's just require gfortran. It's available on all platforms and there are no problems on Windows with running the compiled code, so long as it's in executable form. Then we just have to write a makefile, which should be fairly simple, given that compilation is a one-liner. We then call `make` using `subprocess` in our setup.py.

**Comment by barkls** [comment]
That sounds like a good solution. What is the advantage to calling make
over gfortran directly?

**Comment by barkls** [comment]
@vnmanoharan I can try to implement this, but I have no experience with writing makefiles. It looks like a makefile is useful when there are multiple components to compile, and we only want to recompile those with dependencies on files that have changed. I don't think that describes our scenario. Are there other reasons to use ``make`` rather than ``gfortran`` when calling ``subprocess`` from ``setup.py``?

**Comment by vnmanoharan** [comment]
At the moment there is no real difference between calling `gfortran` from `setup.py` and calling `make`, but developers will appreciate a makefile, and if we eventually write a `configure` script, `setup.py` will have to call both `configure` and `make`.

**Comment by barkls** [comment]
Axisymmetric T-matrix code is implemented in HoloPy as of 3.1 release.
